File: IQ/Classes/Manipulation/ManipulationMethod.vb
Public Class ManipulationMethod
    Property SourceBranchId As Integer? = New Nullable(Of Integer)
    Property TargetBranchId As Integer? = New Nullable(Of Integer)
    Property SourcePath As String
    Property TargetPath As String
    Property LoginId As UInt64
    Property AuditId As Integer?
    Friend errormessages As List(Of String) = New List(Of String)()

    ReadOnly Property TargetBranch As clsBranch
        Get
            If TargetBranchId Is Nothing Then
                If TargetPath Is Nothing Then
                    Return Nothing
                Else
                    Return iq.Branches(CInt(TargetPath.Split(".")(TargetPath.Split(".").Length - 1)))
                End If
            Else
                Return iq.Branches(TargetBranchId)
            End If
        End Get
    End Property
    ReadOnly Property SourceBranch As clsBranch
        Get
            If SourceBranchId Is Nothing Then
                If SourcePath Is Nothing Then
                    Return Nothing
                Else
                    Return iq.Branches(CInt(SourcePath.Split(".")(SourcePath.Split(".").Length - 1)))
                End If
            Else
                Return iq.Branches(SourceBranchId)
            End If
        End Get
    End Property


    Function PerformAction() As String

    End Function

    Function UndoAction() As String
        AuditLog.Instance.MarkUndone(AuditId, LoginId, "Undo" + Me.GetType().Name, errormessages, "") 'Add referer?
    End Function

End Class

File: IQ/Classes/Manipulation/ManipulationMethod.cs
public class ManipulationMethod
{
    public ManipulationMethod()
    {
        // VBConversions Note: Non-static class variable initialization is below.  Class variables cannot be initially assigned non-static values in C#.
        SourceBranchId = new Nullable<int>();
        TargetBranchId = new Nullable<int>();
        errormessages = new List<string>();

    }
    public int? SourceBranchId { get; set; }
    public int? TargetBranchId { get; set; }
    public string SourcePath { get; set; }
    public string TargetPath { get; set; }
    public UInt64 LoginId { get; set; }
    public int? AuditId { get; set; }
    internal List<string> errormessages; // VBConversions Note: Initial value cannot be assigned here since it is non-static.  Assignment has been moved to the class constructors.

    public clsBranch TargetBranch
    {
        get
        {
            if (TargetBranchId == null)
            {
                if (TargetPath == null)
                {
                    return null;
                }
                else
                {
                    return iq.Branches(System.Convert.ToInt32(TargetPath.Split('.')[TargetPath.Split('.').Length - 1]));
                }
            }
            else
            {
                return iq.Branches(TargetBranchId);
            }
        }
    }
    public clsBranch SourceBranch
    {
        get
        {
            if (SourceBranchId == null)
            {
                if (SourcePath == null)
                {
                    return null;
                }
                else
                {
                    return iq.Branches(System.Convert.ToInt32(SourcePath.Split('.')[SourcePath.Split('.').Length - 1]));
                }
            }
            else
            {
                return iq.Branches(SourceBranchId);
            }
        }
    }


    public string PerformAction()
    {

    }

    public string UndoAction()
    {
        AuditLog.Instance.MarkUndone(AuditId, LoginId, "Undo" + this.GetType().Name, errormessages, ""); //Add referer
    }

}